tests/qmctl-test/scripts/qmctl_cp_bidirectional.sh

```diff
@@ -0,0 +1,211 @@
+#!/bin/bash
+
+# Exit immediately on any error
+set -e
+
+# shellcheck disable=SC1091
+. ../e2e/lib/utils
+
+QMCTL_SCRIPT="../../tools/qmctl/qmctl"
+
+info_message "Starting comprehensive qmctl copy tests..."
+
+# Check if container is available
+if ! python3 "$QMCTL_SCRIPT" exec echo "container_check" > /dev/null 2>&1; then
+    fail_message "QM container is not available - copy tests cannot run"
+    echo "Please ensure QM container is running with: sudo systemctl start qm"
+    exit 1
+fi
+
+info_message "QM container is available - running comprehensive copy tests"
+
+# Test 0: Basic copy test (from original qmctl_cp.sh)
+info_message "Testing: Basic copy test (host to container)"
+python3 "$QMCTL_SCRIPT" cp ./files/file-to-copy.txt qm:/tmp
+
+expected_output_file="./files/file-to-check-cp.txt"
+actual_output_temp_file=$(mktemp)
+
+python3 "$QMCTL_SCRIPT" exec cat /tmp/file-to-copy.txt > "$actual_output_temp_file"
+
+if diff "$actual_output_temp_file" "$expected_output_file" > /dev/null; then
+    pass_message "Basic copy - File content matches expected"
+else
+    fail_message "Basic copy - File content does NOT match expected"
+    echo "  Expected: $expected_output_file"
+    echo "  Actual output saved to: $actual_output_temp_file"
+    exit 1
+fi
+rm -f "$actual_output_temp_file"
+
+# Clean up the basic test file
+python3 "$QMCTL_SCRIPT" exec bash -c "rm -f /tmp/file-to-copy.txt && echo 'cleaned up basic test file'" >/dev/null 2>&1 || true
+
+# Function to create a unique test file with specific content
+create_test_file() {
+    local filepath="$1"
+    local content="$2"
+
+    echo "$content" > "$filepath"
+    if [ ! -f "$filepath" ]; then
+        echo "ERROR: Failed to create test file $filepath"
+        return 1
+    fi
+}
+
+# Helper function to verify file content at destination
+verify_copy_content() {
+    local test_name="$1"
+    local dst_path="$2"
+    local expected_content="$3"
+
+    local actual_content=""
+    local actual_dst_path="$dst_path"
+
+    # Handle directory destinations - construct full path
+    if [[ "$dst_path" == */ ]]; then
+        local src_filename
+        src_filename=$(basename "$4")  # $4 is the source path
+        actual_dst_path="${dst_path}${src_filename}"
+    fi
+
+    if [[ "$actual_dst_path" == qm:* ]]; then
+        # Destination is in container - use exec to read content
+        local container_path="${actual_dst_path#qm:}"
+        if actual_content=$(python3 "$QMCTL_SCRIPT" exec cat "$container_path" 2>/dev/null); then
+            # Success - continue with verification
+            :
+        else
+            fail_message "$test_name - Could not read destination file in container"
+            echo "  Tried to read: $container_path"
+            exit 1
+        fi
+
+        # Remove 'output: ' prefix that qmctl adds
+        if [[ "$actual_content" == "output: "* ]]; then
```

Comment:
do we need this if? 

how about just add: actual_content="${actual_content#output: }" ? If output is do something else, do nothing.

Reply:
Changed